packages/fether-react/src/contexts/TokensContext.js

```diff
@@ -0,0 +1,129 @@
+
+import ethereumIcon from '../assets/img/tokens/ethereum.png';
+import LS_PREFIX from '../stores/utils/lsPrefix';
+
+import Debug from '../utils/debug';
+
+const debug = Debug('TokensContext');
+
+const LS_KEY = `${LS_PREFIX}::tokens`;
+
+const DEFAULT_TOKENS = {
+  ETH: {
+    address: 'ETH',
+    decimals: 18,
+    logo: ethereumIcon,
+    name: 'Ether',
+    symbol: 'ETH'
+  }
+};
+
+@subscribe('tokens')
+class Init extends Component {
+  constructor (props) {
+    super(props);
+
+    chainName$({ withoutLoading: true }).subscribe(chainName => {
+      props.setChainName(chainName);
+      props.fetchTokensFromLS();
+    });
+  }
+
+  render () {
+    return this.props.children;
+  }
+}
+
+class WithProvider extends Component {
+  tokensStore = null;
+
+  // (won't update if props.accountAddress changes)
+  constructor (props) {
+    super(props);
+    const { accountAddress } = props;
+
+    // Cannot put the function in the store or else it's considered an action
+    // and must return the new state
+    const getLsKey = () =>
+      // We have one key per chain per account, in this format:
+      // __paritylight::tokens::0x123::kovan
+      `${LS_KEY}::${this.tokensStore.accountAddress}::${
+        this.tokensStore.chainName
+      }`;
+
+    this.tokensStore = {
+      accountAddress,
+      chainName: null,
+      tokens: {},
+      setChainName: chainName => ({ chainName }),
+      fetchTokensFromLS: () => state => {
+        debug(
+          `Fetching tokens for account ${state.accountAddress} and chain ${
+            state.chainName
+          }`
+        );
+        const lsKey = getLsKey();
+        const tokensFromLS = store.get(lsKey);
+
+        if (tokensFromLS) {
+          debug(`Got tokens from localStorage`, tokensFromLS);
+          return { tokens: tokensFromLS };
+        } else {
+          // If there's nothing in the localStorage, we add by default only
+          // Ethereum. We consider Ethereum as a token, with address 'ETH'
+          debug(`No tokens in localStorage; setting default.`);
+          store.set(lsKey, DEFAULT_TOKENS);
+          return { tokens: DEFAULT_TOKENS };
+        }
+      },
+      addToken: (address, token) => state => {
+        const newTokens = { ...state.tokens, [address]: token };
+        store.set(getLsKey(), newTokens);
+        return { tokens: newTokens };
+      },
+      removeToken: address => state => {
+        const { [address]: _, ...newTokens } = state.tokens;
+        store.set(getLsKey(), newTokens);
+        return { tokens: newTokens };
+      }
+    };
+  }
+
+  render () {
+    const { Component, ...propsRest } = this.props;
+
+    return (
+      <Provider id='tokens' {...this.tokensStore}>
+        <Init>
+          <Component {...propsRest} />
+        </Init>
+      </Provider>
+    );
+  }
+}
+
+// Decorated class needs to have an accountAddress prop
+export const provideTokens = Component =>
+  withProps({ Component })(WithProvider);
```

Comment:
By reading react-contextual readme, I don't think provideTokens should be a HOC.

I think it'll be simpler to have in Tokens.js (the top-level component that needs the token context), add `<Provider id='tokens' {...tokensStore}>` in its render method. In this case we're sure that there's only one Provider.

And then consumers are defined with HOC.

Reply:
hmm but then how do we initialize tokensStore with the account address? and how can we subscribe to `chainName$` and update the store accordingly? (i.e. what `<Init>` does currently)

Reply:
Exactly, I quickly looked into react-contextual, didn't find anything promising.

Would `<Provider id='tokens' accountAddres={...} chainName={...} {...tokensStore}>` work? And we subscribe Tokens.js to light.js to get those props.

Reply:
ah, yes, I think it would work!

so Tokens.js would be `<Provider id='tokens' accountAddress={...} chainName={...} {...tokensStore}><Subscribe id='tokens'>[...]</Subscribe></Provider>`

but when we're subscribing to the chainName$ observable, we need to call a context prop from inside Subscribe (`props.setChainName(chainName);` -- we cannot do `tokensStore.setChainName(chainName)` because then the return value of the action isn't evaluated by react-contextual), so I think we would need another component... doesn't seem straightforward to me

Reply:
No, the setChainName method disappears completely. Only Tokens.js would subscribe to chainName from light.js, and then inside the Provider's store functions `state.chainName` would be defined, no need to setChainName at all.

Although I'm not 100% sure of what I'm saying, just thinking that it'd be nice if react-contextual worked like this.

Reply:
ah, I understand
but we still need to call the store function `fetchTokens`, and this needs to be done from within the Subscribe component using its props, in order for the return value of the function (action) to be evaluated by react-contextual